### ISSUE TITLE
TCP liveness checks through Envoy

### ISSUE THREAD

**Issue by freddygv** [envoyproxy/envoy#7783]
Is there a way for a service fronted by Envoy to be checked for liveness with a TCP connection? (Assuming the service cannot be reached directly without going through the proxy)

It seems that Envoy will accept any TCP connection, so the liveness check will always pass. 

Envoy can set up active TCP checks for connection only as shown [here](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/health_check.proto#core-healthcheck-tcphealthcheck). I would like that behavior but with an application on a different pod/node running the check.

**Comment by ishustava** [comment]
Could we re-open this issue? This seems like a good question. 

Right now if I'm using TCP checks for a service fronted by Envoy and that cannot be reached directly, the TCP check will pass regardless of the port I'm checking because it always has to go to the Envoy's port. It feels like something that would be good to have so that if I do have a TCP service, I can check that it's running on its port.

**Comment by mattklein123** [comment]
Can you describe how you want this to work? Something like the HTTP health check filter but for listeners?

**Comment by lambdai** [comment]
> Could we re-open this issue? This seems like a good question.
> 
> Right now if I'm using TCP checks for a service fronted by Envoy and that cannot be reached directly, the TCP check will pass regardless of the port I'm checking because it always has to go to the Envoy's port. It feels like something that would be good to have so that if I do have a TCP service, I can check that it's running on its port.

@ishustava Do you mean envoy start listening on certain port too early? I am aware of some cases, including listener config that uses reuse_port == false. 
Envoy doesn't listen on all port so I want to understand why the tcp check was failing when hitting your Envoy instance.

**Comment by freddygv** [comment]
@lambdai The TCP check isn't failing. The issue is that it always passes as long as Envoy is listening, which is independent from whether the local application is listening.

@mattklein123 That sounds about right from looking at this documentation: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/health_checking#http-health-checking-filter. 

For inbound TCP checks the current behavior seems roughly equivalent to "No pass through" in the HTTP health check filter. Being able to open a connection to Envoy's listener isn't reflective of whether the local application is listening for TCP connections. 

Ideally, if the local application isn't running or otherwise not listening, health checkers should not be able to open a connection to Envoy's inbound listener. 

The use-case here is when there's some service, like the Kubernetes kubelet, that is configured to run liveness probes against TCP applications fronted by Envoy (and cannot dial them directly).

**Comment by mattklein123** [comment]
Yeah this makes sense and would be a nice feature. I will mark help wanted.

**Comment by lambdai** [comment]
@mattklein123 I think this is reasonable from this use case. It arguable when there are more than 2 network filter chains at the same tcp listener. 
Suppose one tcp proxy at this listener reports not healthy while the other tcp proxy at the same listener but owned by another network filter chain reports healthy, it's not reasonable to stop listening by failing the tcp shake.

**Comment by mattklein123** [comment]
Yeah I agree it needs more thinking and design.

**Comment by sabyasachiroy** [comment]
What is the current status of this? I understand that there might be different expected behaviour, i.e., if local application is not up then fail the check versus not. But the passthrough check seems a common enough use-case to allow it via some config and let the users of envoy decide which one is applicable.
